src/vm/loaderallocator.cpp

```diff
@@ -2046,4 +2053,126 @@ BOOL LoaderAllocator::InsertComInteropData(MethodTable* pMT, InteropMethodTableD
 
 #endif // FEATURE_COMINTEROP
 
+
+void LoaderAllocator::AddUnmanagedImageToCache(LPCWSTR libraryName, HMODULE hMod)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(libraryName));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+    if (libraryName)
+    {
+        m_UnmanagedCache.InsertEntry(libraryName, hMod);
+    }
+    return ;
+}
+
+HMODULE LoaderAllocator::FindUnmanagedImageInCache(LPCWSTR libraryName)
+{
+    CONTRACT(HMODULE)
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(libraryName,NULL_OK));
+        POSTCONDITION(CheckPointer(RETVAL,NULL_OK));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACT_END;
+    if(libraryName == NULL) RETURN NULL;
+
+    RETURN m_UnmanagedCache.LookupEntry(libraryName);
+}
+
+/* static */
+BOOL UnmanagedImageCache::CompareBindingSpec(UPTR spec1, UPTR spec2)
+{
+    WRAPPER_NO_CONTRACT;
+
+    LPCWSTR libraryName = (LPCWSTR) (spec1 << 1);
```

Comment:
Can you add a comment for why we can apply a right shift by 1 to `spec1`? What does this remove or give us?